Interpretable encoding scheme from dataset dimensions to data arrays.

alxmrs opened the pull request:
Here, I've introduced dataclass that encodes and decodes a (mostly) interpretable `xarray.DataArray` of floats. These are manually tested with a fuzzer (hypothesis). These tests are turned off by default due to the choice to encode the numbers as significant digits in floats due to the tendency for rounding errors in `np.float64` values. For our purposes of making debuggable test data, these should be decent.

Fixes #79.

alxmrs committed:
Here, I've introduced dataclass that encodes and decodes a (mostly) interpretable xarray.DataArray of floats. These are manually tested with a fuzzer (hypothesis). These tests are turned off by default due to the choice to encode the numbers as significant digits in floats due to the tendency for rounding errors in np.float64 values. For our purposes of making debuggable test data, these should be decent.

jder commented on tests/conftest.py:
```diff
@@ -1,30 +1,183 @@
+            and self.start_day == other.start_day
+        )
+
+    def parse_cftime_range(self, time_range: xr.CFTimeIndex) -> None:
```
maybe `set_time_range`?

jder commented on tests/conftest.py:
```diff
@@ -1,30 +1,183 @@
+class DataSourceDims:
+    """Dimension metadata to produce interpretable `xarray.DataArray`s.
+
+    Each float in the `xarray.DataArray` has the following scheme:
```
```suggestion
    Each float in the encoded `xarray.DataArray` has the following scheme:
```

jder commented on tests/conftest.py:
```diff
@@ -1,30 +1,183 @@
+        self.start_day = time_range[0]
+        timedeltas = [date - self.start_day for date in time_range]
+        self.days_since_start = np.array(
+            [delta.total_seconds() / (24 * 3600) for delta in timedeltas],
```
I see there's a `cftime.date2num` -- would that make this easier?

alxmrs replied:
TIL, thanks, this made it better.

jder commented on tests/conftest.py:
```diff
@@ -1,30 +1,183 @@
+        time = np.array(
+            [
+                self.start_day + cftime._cftime.timedelta(days=int(d))
+                for d in self.days_since_start
```
Same here, would `cftime.num2date` be easier?

If not, it looks like this is a private alias for `datetime.timedelta`, so let's use that instead?

alxmrs replied:
That idea is great, looks much cleaner.

jder commented on tests/test_datasets.py:
```diff
@@ -1,8 +1,15 @@
+from hypothesis import example, given, settings
+from hypothesis import strategies as st
+from hypothesis.extra.numpy import arrays
+from numpy._typing import NDArray
```
```suggestion
from numpy.typing import NDArray
```

jder commented on tests/test_datasets.py:
```diff
@@ -66,24 +73,98 @@ def extract_sample_arrays(td: TrainData) -> tuple[np.ndarray, np.ndarray]:
+# (2025-03-03) due to floating point rounding woes. In the future,
+# there may be merit to an encoding scheme made of mangled complex
+# numbers. Not worth pursuing right now...
+@pytest.mark.manual
```
Maybe `@pytest.mark.skip` until we want to pursue it?

jder commented on tests/conftest.py:
```diff
@@ -1,30 +1,183 @@
+    )
+
+    def __post_init__(self):
+        self.lat = _trunc(self.lat.astype(np.float64), 2) % 100
```
🔴 If the passed `lat` value is negative, this will make it positive (mod 100) and so potentially overlap with positive `lat` values. Maybe we want this to be % 200 or something so the ranges are distinct? 

In general manipulating the values in `__post_init__` makes it a little hard to believe the test since eg setting them all to zero would cause the test to pass. Maybe it would be better to assert the types and ranges here rather than coercing them?

alxmrs replied:
Thanks, this makes me reflect on the spec. I realized there is an implicit constraint -- only the lat values can be negative. I've now added more asserts and made this constraint explicit. I've also fixed the bug you found wrt to mod and sign.

jder commented on tests/conftest.py:
```diff
@@ -1,30 +1,183 @@
+
+        days_since_start = ((tim_dim * 1000) % (tim_sign * 1000)).astype(np.int32)
+        lat = (lat_dim // (lat_sign * 10_000)) / 100
+        lng = lng_dim / 10 % (lng_sign * 1000)
```
I don't understand the goal of this sign-related logic. It looks like these will always produce positive days/lat/lng values. Is that the goal? When do we expect the encoded floats to be negative? (Can we have them always be positive and remove this logic?)

jder reviewed (changes requested):
Just one blocking thing from me and a few suggestions/questions. If you would prefer to treat this as a hash instead and avoid having to resolve these, I'm ok with that too. (Let's just finish that conversation you started in slack about it.)

alxmrs committed:
Apply suggestions from code review

Thanks @jder.

Co-authored-by: Jesse Rusak <jesse@openathena.ai>

alxmrs committed:
Implemented feedback from @jder. And!! I got the hypothesis round-trip tests to pass!!! This required changing the array equality to allow 0.1 error tolerance. This change includes:

- Stricter value enforcement in dataclass __post_init__ via asserts.
- changing equality to use float tolerance (makes hypothesis tests pass).
- Explicitly making all values non-negative besides lat.
- using cftime `num2date` and `date2num` instead of rolling my own.
- Adding data var index asserts; adding another digit for data_var_index.
- Adding a helpful explainer about signs and modulus.

alxmrs committed:
Self-review: added docs.

alxmrs committed:
rm print stmts.

alxmrs committed:
Longer deadline for CI.

jder commented on tests/conftest.py:
```diff
@@ -1,29 +1,203 @@
+        # All other values in the encoding must be non-negative.
+        self.lng = _trunc(self.lng.astype(np.float64), 1) % 1000
+        if np.any(self.lng < 0.0) or np.any(self.lng > 360.0):
+            raise ValueError("lng values must be between 0 and 360 degrees.")
```
Should this test be on the self.lat/self.lng values *before* we do the trunc and mod?

alxmrs replied:
I originally had that and my conclusion is -- _no_. The reason: every once in a while, we hit an error where [360.] as specified input gets interpreted as [360.09999999] and breaks the assert. I though the same thing at first, too.

jder replied:
Hrm, I don't understand how that's possible given the constraints passed to hypothesis that has a max of 360.

alxmrs replied:
I manually added the edge case example here, and that's what breaks this: https://github.com/suryadheeshjith/Ocean_Emulator/pull/88/files#diff-3b63cb613626e7d787b8e63ced54267d5d18966f5fbc3795812573b4e50a942dR116

And, I think that this is a regular type of thing to do / should be accommodated.

jder reviewed (approved):
Awesome!

alxmrs committed:
again, a longer deadline for CI.
